File: solutions/python/08/monoid_seg/main.py
class MonoidSeg:
    def __init__(self, data, op, e):
        self.n = len(data)
        self.op = op
        self.e = e
        size = 1
        while size < self.n:
            size <<= 1
        self.size = size
        self.dat = [e] * (2 * size)
        for i, x in enumerate(data):
            self.dat[size + i] = x
        for i in range(size - 1, 0, -1):
            self.dat[i] = op(self.dat[i * 2], self.dat[i * 2 + 1])

    def update(self, i, x):
        i += self.size
        self.dat[i] = x
        while i > 1:
            i //= 2
            self.dat[i] = self.op(self.dat[i * 2], self.dat[i * 2 + 1])

    def query(self, l, r):
        l += self.size
        r += self.size
        left = right = self.e
        while l < r:
            if l & 1:
                left = self.op(left, self.dat[l])
                l += 1
            if r & 1:
                r -= 1
                right = self.op(self.dat[r], right)
            l //= 2
            r //= 2
        return self.op(left, right)


def main() -> None:
    data = [1, 3, 2, 5, 4]
    mx = MonoidSeg(data, max, -10**30)
    sm = MonoidSeg(data, lambda a, b: a + b, 0)
    print(mx.query(1, 4))
    print(sm.query(0, 5))
    mx.update(2, 10)
    print(mx.query(0, 5))


if __name__ == "__main__":
    main()

File: solutions/go/08/monoid_seg/main.go
package main

import "fmt"

type monoidSeg struct {
	size int
	dat  []int
	op   func(int, int) int
	e    int
}

func newMonoidSeg(data []int, op func(int, int) int, e int) *monoidSeg {
	n := len(data)
	size := 1
	for size < n {
		size <<= 1
	}
	s := &monoidSeg{size: size, op: op, e: e, dat: make([]int, 2*size)}
	for i, x := range data {
		s.dat[size+i] = x
	}
	for i := size - 1; i > 0; i-- {
		s.dat[i] = op(s.dat[i*2], s.dat[i*2+1])
	}
	return s
}

func (s *monoidSeg) query(l, r int) int {
	l += s.size
	r += s.size
	left, right := s.e, s.e
	for l < r {
		if l&1 == 1 {
			left = s.op(left, s.dat[l])
			l++
		}
		if r&1 == 1 {
			r--
			right = s.op(s.dat[r], right)
		}
		l >>= 1
		r >>= 1
	}
	return s.op(left, right)
}

func main() {
	data := []int{1, 3, 2, 5, 4}
	mx := newMonoidSeg(data, func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}, -1_000_000_000)
	sum := newMonoidSeg(data, func(a, b int) int { return a + b }, 0)
	fmt.Println(mx.query(1, 4))
	fmt.Println(sum.query(0, 5))
}
